pstoreds/addr_book.go

```diff
@@ -169,42 +190,54 @@ func (mgr *dsAddrBook) setAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Durati
 			mgr.subsManager.BroadcastAddr(p, addrs[i])
 		}
 	}
-
-	// Force update TTLs only if TTL reset was requested; otherwise
-	// insert the appropriate TTL entries if they don't already exist.
-	if ttlReset {
-		mgr.ttlManager.setTTLs(keys, ttl)
-	} else {
-		mgr.ttlManager.insertOrExtendTTLs(keys, ttl)
-	}
-
 	return nil
 }
 
 // dbInsert performs a transactional insert of the provided keys and values.
-func (mgr *dsAddrBook) dbInsert(keys []ds.Key, addrs []ma.Multiaddr) ([]bool, error) {
+func (mgr *dsAddrBook) dbInsert(keys []ds.Key, addrs []ma.Multiaddr, ttl time.Duration, mode ttlWriteMode) ([]bool, error) {
 	var (
 		err     error
 		existed = make([]bool, len(keys))
+		exp     = time.Now().Add(ttl)
+		ttlB    = make([]byte, 8)
 	)
 
+	binary.LittleEndian.PutUint64(ttlB, uint64(ttl))
+
 	txn := mgr.ds.NewTransaction(false)
 	defer txn.Discard()
 
+	ttltxn := txn.(ds.TTLDatastore)
 	for i, key := range keys {
 		// Check if the key existed previously.
-		if existed[i], err = txn.Has(key); err != nil {
+		if existed[i], err = ttltxn.Has(key); err != nil {
 			log.Errorf("transaction failed and aborted while checking key existence: %s, cause: %v", key.String(), err)
 			return nil, err
 		}
 
-		// The key embeds a hash of the value, so if it existed, we can safely skip the insert.
+		// The key embeds a hash of the value, so if it existed, we can safely skip the insert and
+		// just update the TTL.
 		if existed[i] {
+			switch mode {
+			case ttlOverride:
+				err = ttltxn.SetTTL(key, ttl)
+			case ttlExtend:
+				var curr time.Time
+				if curr, err = ttltxn.GetExpiration(key); err == nil && exp.After(curr) {
+					err = ttltxn.SetTTL(key, ttl)
+				}
+			}
+			if err != nil {
+				// mode will be printed as an int
+				log.Errorf("failed while updating the ttl for key: %s, mode: %v, cause: %v", key.String(), mode, err)
+				return nil, err
+			}
 			continue
 		}
 
-		// Attempt to add the key.
-		if err = txn.Put(key, addrs[i].Bytes()); err != nil {
+		// format: bytes(ttl) || bytes(multiaddr)
+		value := append(ttlB, addrs[i].Bytes()...)
```

Comment:
let's make a record type, i.e.

```golang
type AddrRecord struct {
  Addrs []ma.Multiaddr
  TTL time.Duration
}
```

and either hand write serializer/deserializer or use gob (simplest). that way we can isolate serialization/deserialization